File: care/emr/reports/context_builder/data_points/encounter.py
from types import SimpleNamespace

from django_filters import rest_framework as filters

from care.emr.models.encounter import Encounter, EncounterOrganization
from care.emr.models.tag_config import TagConfig
from care.emr.reports.context_builder.data_point_registry import DataPointRegistry
from care.emr.reports.context_builder.data_points.allergy_intolerance import (
    AllergyIntoleranceContextBuilder,
)
from care.emr.reports.context_builder.data_points.base import (
    Field,
    QuerysetContextBuilder,
    SingleObjectContextBuilder,
)
from care.emr.reports.context_builder.data_points.diagnosis import (
    DiagnosisContextBuilder,
)
from care.emr.reports.context_builder.data_points.diagnostic_report import (
    DiagnosticReportContextBuilder,
)
from care.emr.reports.context_builder.data_points.facility import FacilityContextBuilder
from care.emr.reports.context_builder.data_points.medication import (
    MedicationPrescriptionContextBuilder,
)
from care.emr.reports.context_builder.data_points.patient import (
    IdentifiersContextBuilder,
    PatientMinimumContextBuilder,
    PatientTagContextBuilder,
    TagFilter,
)
from care.emr.reports.context_builder.data_points.questionnaire import (
    QuestionnaireContextBuilder,
)
from care.emr.reports.context_builder.data_points.service_request import (
    ServiceRequestBaseContextBuilder,
)
from care.emr.reports.context_builder.data_points.symptom import SymptomsContextBuilder
from care.emr.reports.context_builder.data_points.user import SingleUserIdContextBuilder

STATUS_DISPLAY = {
    "planned": "Planned",
    "in_progress": "In Progress",
    "completed": "Completed",
    "cancelled": "Cancelled",
    "entered_in_error": "Entered in Error",
}

ENCOUNTER_CLASS_DISPLAY = {
    "imp": "Inpatient",
    "amb": "Outpatient",
    "obsenc": "Observation",
    "emer": "Emergency",
    "vr": "Virtual",
    "hh": "Home Health",
}

ENCOUNTER_PRIORITY_DISPLAY = {
    "ASAP": "ASAP",
    "callback_results": "Callback Results",
    "callback_for_scheduling": "Callback for Scheduling",
    "elective": "Elective",
    "emergency": "Emergency",
    "preop": "Preop",
    "as_needed": "As Needed",
    "routine": "Routine",
    "rush_reporting": "Rush Reporting",
    "stat": "Stat",
    "timing_critical": "Timing Critical",
    "use_as_directed": "Use as Directed",
    "urgent": "Urgent",
}


class EncounterCareTeamContextBuilder(QuerysetContextBuilder):
    def get_context(self):
        return self.parent_context.care_team

    user = Field(
        display="User",
        target_context=SingleUserIdContextBuilder,
        preview_value="",
        description="User who is part of the encounter care team",
    )
    role = Field(
        display="Role",
        preview_value="Primary care physician",
        mapping=lambda c: c.role.get("display")
        if c.role and c.role.get("display")
        else "",
        description="Role of the user in the encounter care team",
    )

    def __iter__(self):
        if self.is_preview:
            return iter(self.__class__(is_preview=True) for c in range(3))
        return iter(
            self.__class__(context=SimpleNamespace(user=c["user_id"], role=c["role"]))
            for c in self.context
        )


class EncounterPatientFacilityTagContextBuilder(PatientTagContextBuilder):
    filterset_class = TagFilter
    __filterset_backends__ = [filters.DjangoFilterBackend]

    def get_context(self):
        return TagConfig.objects.filter(
            id__in=self.parent_context.patient.facility_tags.get(
                str(self.parent_context.facility.id), []
            )
        )


class EncounterTagContextBuilder(PatientTagContextBuilder):
    filterset_class = TagFilter
    __filterset_backends__ = [filters.DjangoFilterBackend]

    def get_context(self):
        return TagConfig.objects.filter(id__in=self.parent_context.tags or [])


class EncounterFacilityLocationContextBuilder(SingleObjectContextBuilder):
    def get_context(self):
        return getattr(self.parent_context, self.parent_attribute)

    name = Field(
        display="Location Name",
        preview_value="Ward A",
        description="Name of the facility location",
    )


class EncounterOrganizationsContextBuilder(QuerysetContextBuilder):
    def get_context(self):
        return EncounterOrganization.objects.filter(encounter_id=self.parent_context.id)

    organization = Field(
        display="Organization",
        preview_value="",
        mapping=lambda o: o.organization.name if o.organization else "",
        description="Organization associated with the encounter",
    )


HOSPITALIZATION_ADMIT_SOURCE_DISPLAY = {
    "hosp_trans": "Transferred from other hospital",
    "emd": "From accident/emergency department",
    "outp": "From outpatient department",
    "born": "Born in hospital",
    "gp": "General Practitioner referral",
    "mp": "Medical Practitioner/physician referral",
    "nursing": "From nursing home",
    "psych": "From psychiatric hospital",
    "rehab": "From rehabilitation facility",
    "other": "Other",
}

HOSPITALIZATION_DISCHARGE_DISPOSITION_DISPLAY = {
    "home": "Home",
    "alt_home": "Alternate Home",
    "other_hcf": "Other Health Care Facility",
    "hosp": "Hospital",
    "long": "Long-term Care Facility",
    "aadvice": "Against Medical Advice",
    "exp": "Expired",
    "psy": "Psychiatric Hospital",
    "rehab": "Rehabilitation Facility",
    "snf": "Skilled Nursing Facility",
    "oth": "Other",
}

HOSPITALIZATION_DIET_PREFERENCE_DISPLAY = {
    "vegetarian": "Vegetarian",
    "dairy_free": "Dairy Free",
    "nut_free": "Nut Free",
    "gluten_free": "Gluten Free",
    "vegan": "Vegan",
    "halal": "Halal",
    "kosher": "Kosher",
    "none": "None",
}


class EncounterHospitalizationContextBuilder(SingleObjectContextBuilder):
    def get_context(self):
        return self.parent_context.hospitalization or {}

    re_admission = Field(
        display="Re-admission",
        mapping=lambda h: h.get("re_admission"),
        preview_value="False",
        description="Whether the encounter is a re-admission",
    )
    admit_source = Field(
        display="Admit Source",
        mapping=lambda h: HOSPITALIZATION_ADMIT_SOURCE_DISPLAY.get(
            h.get("admit_source")
        )
        if h.get("admit_source")
        else "",
        preview_value="From accident/emergency department",
        description="Source of admission for the encounter",
    )
    discharge_disposition = Field(
        display="Discharge Disposition",
        mapping=lambda h: HOSPITALIZATION_DISCHARGE_DISPOSITION_DISPLAY.get(
            h.get("discharge_disposition")
        )
        if h.get("discharge_disposition")
        else "",
        preview_value="",
        description="Disposition of discharge for the encounter",
    )
    diet_preference = Field(
        display="Diet Preference",
        mapping=lambda h: HOSPITALIZATION_DIET_PREFERENCE_DISPLAY.get(
            h.get("diet_preference")
        )
        if h.get("diet_preference")
        else "",
        preview_value="Dairy Free",
        description="Diet preference for the encounter",
    )


class BaseEncounterReportContext(SingleObjectContextBuilder):
    status = Field(
        display="Encounter Status",
        mapping=lambda e: STATUS_DISPLAY.get(
            e.status, e.status.title() if e.status else ""
        ),
        preview_value="In Progress",
        description="Current status of the encounter",
    )

    encounter_class = Field(
        display="Encounter Class",
        mapping=lambda e: ENCOUNTER_CLASS_DISPLAY.get(
            e.encounter_class, e.encounter_class.title() if e.encounter_class else ""
        ),
        preview_value="Outpatient",
        description="Classification of the encounter",
    )
    care_team = Field(
        target_context=EncounterCareTeamContextBuilder,
        display="Care Team",
        preview_value="",
        description="Care team of the encounter",
    )

    current_location = Field(
        display="Current Location",
        target_context=EncounterFacilityLocationContextBuilder,
        preview_value="",
        description="Current location within the facility for the encounter",
    )

    start_time = Field(
        display="Encounter Start Time",
        mapping=lambda e: e.period.get("start") if e.period else None,
        preview_value="2026-01-12T10:01:45.088000Z",
        description="Start time of the encounter",
    )
    end_time = Field(
        display="Encounter End Time",
        mapping=lambda e: e.period.get("end")
        if e.period and e.period.get("end")
        else "Ongoing",
        preview_value="2026-01-12T10:01:45.088000Z",
        description="End time of the encounter",
    )
    organizations = Field(
        display="Associated Organizations",
        target_context=EncounterOrganizationsContextBuilder,
        preview_value="",
        description="Organizations associated with the encounter",
    )
    discharge_summary_advice = Field(
        display="Discharge Summary Advice",
        mapping="discharge_summary_advice",
        preview_value="Patient is advised to follow up in 2 weeks.",
        description="Discharge summary advice for the encounter",
    )
    patient_facility_tags = Field(
        target_context=EncounterPatientFacilityTagContextBuilder,
        display="Patient Facility Tags",
        preview_value="",
        description="Facility-specific tags associated with the patient for the encounter",
    )
    encounter_tags = Field(
        target_context=EncounterTagContextBuilder,
        display="Encounter Tags",
        preview_value="",
        description="Tags associated with the encounter",
    )
    priority = Field(
        display="Priority",
        mapping=lambda e: ENCOUNTER_PRIORITY_DISPLAY.get(
            e.priority, e.priority.title() if e.priority else ""
        ),
        preview_value="ASAP",
        description="Priority of the encounter",
    )
    hospitalization = Field(
        target_context=EncounterHospitalizationContextBuilder,
        display="Hospitalization",
        preview_value="",
        description="Hospitalization of the encounter",
    )
    external_identifier = Field(
        display="External Identifier",
        mapping="external_identifier",
        preview_value="1234567890",
        description="External identifier of the encounter",
    )


class MinimumEncounterReportContext(BaseEncounterReportContext):
    def get_context(self):
        return getattr(self.parent_context, self.parent_attribute)


class PatientFacilityIdentifiersContextBuilder(IdentifiersContextBuilder):
    def get_context(self):
        return self.parent_context.patient.facility_identifiers.get(
            str(self.parent_context.facility.id), []
        )


class EncounterReportContext(BaseEncounterReportContext):
    standalone_context = True
    __slug__ = "encounter_base"
    __associating_model__ = Encounter
    __display_name__ = "Encounter Report"
    __description__ = "Report context for encounter-based reports"
    context_key = "encounter"

    patient = Field(
        display="Patient Details",
        target_context=PatientMinimumContextBuilder,
        preview_value="",
        description="Details of the patient associated with the encounter",
    )

    facility = Field(
        display="Facility Details",
        target_context=FacilityContextBuilder,
        preview_value="",
        description="Details of the facility where the encounter took place",
    )

    diagnostic_reports = Field(
        display="Diagnostic Reports",
        preview_value="",
        description="Diagnostic reports associated with the encounter",
        target_context=DiagnosticReportContextBuilder,
    )
    symptoms = Field(
        target_context=SymptomsContextBuilder,
        display="Symptoms",
        preview_value="",
        description="Symptoms of the encounter",
    )
    allergy_intolerances = Field(
        target_context=AllergyIntoleranceContextBuilder,
        display="Allergy Intolerances",
        preview_value="",
        description="Allergy intolerances of the encounter",
    )
    diagnoses = Field(
        target_context=DiagnosisContextBuilder,
        display="Diagnoses",
        preview_value="",
        description="Diagnoses of the encounter",
    )

    questionnaire_responses = Field(
        target_context=QuestionnaireContextBuilder,
        display="Questionnaire Responses",
        preview_value="",
        description="Questionnaire responses of the encounter",
    )

    medication_prescriptions = Field(
        display="Medication Prescriptions",
        target_context=MedicationPrescriptionContextBuilder,
        preview_value="",
        description="Medication prescriptions of the encounter",
    )

    service_requests = Field(
        display="Service Requests",
        target_context=ServiceRequestBaseContextBuilder,
        preview_value="",
        description="Service requests associated with the encounter",
    )

    facility_identifiers = Field(
        display="Facility Identifiers",
        target_context=PatientFacilityIdentifiersContextBuilder,
        preview_value="",
        description="Facility identifiers associated with the encounter",
    )


DataPointRegistry.register(EncounterReportContext)

File: frontend/src/types/emr/encounter/encounter.ts
import {
  BadgeCheck,
  Ban,
  BedDouble,
  Calendar,
  CirclePause,
  CircleX,
  DoorOpen,
  FileQuestion,
  Home,
  HouseHeart,
  LoaderCircle,
  LucideIcon,
  MonitorSmartphone,
  Siren,
  Stethoscope,
} from "lucide-react";

import { Badge } from "@/components/ui/badge";

import { CareTeamResponse } from "@/types/careTeam/careTeam";
import { PatientListRead, PatientRead } from "@/types/emr/patient/patient";
import { Permissions } from "@/types/emr/permission/permission";
import { TagConfig } from "@/types/emr/tagConfig/tagConfig";
import { FacilityBareMinimum } from "@/types/facility/facility";
import { FacilityOrganizationRead } from "@/types/facilityOrganization/facilityOrganization";
import { LocationAssociationRead } from "@/types/location/association";
import { LocationMinSpec, LocationRead } from "@/types/location/location";
import { AppointmentRead } from "@/types/scheduling/schedule";
import { UserReadMinimal } from "@/types/user/user";

export const ENCOUNTER_ADMIT_SOURCE = [
  "hosp_trans",
  "emd",
  "outp",
  "born",
  "gp",
  "mp",
  "nursing",
  "psych",
  "rehab",
  "other",
] as const;

/**
 * Do not use this constant directly. Use `careConfig.encounterClasses` instead.
 */
export const ENCOUNTER_CLASS = [
  "imp",
  "amb",
  "obsenc",
  "emer",
  "vr",
  "hh",
] as const;

export const ENCOUNTER_DIET_PREFERENCE = [
  "vegetarian",
  "dairy_free",
  "nut_free",
  "gluten_free",
  "vegan",
  "halal",
  "kosher",
  "none",
] as const;

export const ENCOUNTER_DISCHARGE_DISPOSITION = [
  "home",
  "other_hcf",
  "aadvice",
  "exp",
  "alt_home",
  "hosp",
  "long",
  "psy",
  "rehab",
  "snf",
  "oth",
] as const;

export const ENCOUNTER_PRIORITY = [
  "stat",
  "ASAP",
  "emergency",
  "urgent",
  "routine",
  "elective",
  "rush_reporting",
  "timing_critical",
  "callback_results",
  "callback_for_scheduling",
  "preop",
  "as_needed",
  "use_as_directed",
] as const;

export const ENCOUNTER_PRIORITY_COLORS = {
  stat: "destructive",
  ASAP: "yellow",
  emergency: "destructive",
  urgent: "orange",
  routine: "blue",
  elective: "indigo",
  rush_reporting: "orange",
  timing_critical: "yellow",
  callback_results: "green",
  callback_for_scheduling: "purple",
  preop: "pink",
  as_needed: "teal",
  use_as_directed: "indigo",
} as const satisfies Record<
  EncounterPriority,
  React.ComponentProps<typeof Badge>["variant"]
>;

export enum EncounterStatus {
  PLANNED = "planned",
  IN_PROGRESS = "in_progress",
  ON_HOLD = "on_hold",
  DISCHARGED = "discharged",
  COMPLETED = "completed",
  CANCELLED = "cancelled",
  DISCONTINUED = "discontinued",
  ENTERED_IN_ERROR = "entered_in_error",
  UNKNOWN = "unknown",
}

export const ENCOUNTER_STATUS_COLORS = {
  planned: "blue",
  in_progress: "yellow",
  on_hold: "orange",
  discharged: "primary",
  completed: "green",
  cancelled: "destructive",
  discontinued: "destructive",
  entered_in_error: "destructive",
  unknown: "secondary",
} as const satisfies Record<
  EncounterStatus,
  React.ComponentProps<typeof Badge>["variant"]
>;

export const ENCOUNTER_CLASS_ICONS = {
  imp: BedDouble,
  amb: DoorOpen,
  obsenc: Stethoscope,
  emer: Siren,
  vr: MonitorSmartphone,
  hh: HouseHeart,
} as const satisfies Record<EncounterClass, LucideIcon>;

export const ENCOUNTER_STATUS_ICONS = {
  planned: Calendar,
  in_progress: LoaderCircle,
  discharged: Home,
  completed: BadgeCheck,
  cancelled: CircleX,
  on_hold: CirclePause,
  discontinued: CircleX,
  entered_in_error: Ban,
  unknown: FileQuestion,
} as const satisfies Record<EncounterStatus, LucideIcon>;

export const ENCOUNTER_CLASSES_COLORS = {
  imp: "indigo", // Inpatient
  emer: "destructive", // Emergency
  amb: "green", // Outpatient/Ambulatory
  obsenc: "secondary", // Observation
  vr: "secondary", // Virtual
  hh: "teal", // Home Health
} as const satisfies Record<EncounterClass, string>;

export type EncounterAdmitSources = (typeof ENCOUNTER_ADMIT_SOURCE)[number];

export type EncounterClass = (typeof ENCOUNTER_CLASS)[number];

export type EncounterDietPreference =
  (typeof ENCOUNTER_DIET_PREFERENCE)[number];

export type EncounterDischargeDisposition =
  (typeof ENCOUNTER_DISCHARGE_DISPOSITION)[number];

export type EncounterPriority = (typeof ENCOUNTER_PRIORITY)[number];

export type Period = {
  start?: string;
  end?: string;
};

export type Hospitalization = {
  re_admission?: boolean;
  admit_source?: EncounterAdmitSources;
  discharge_disposition?: EncounterDischargeDisposition;
  diet_preference?: EncounterDietPreference;
};

export type History = {
  status: string;
  moved_at: string;
};

export type EncounterClassHistory = {
  history: History[];
};

export type StatusHistory = {
  history: History[];
};

export interface EncounterBase {
  status: EncounterStatus;
  encounter_class: EncounterClass;
  period: Period;
  hospitalization?: Hospitalization | null;
  priority: EncounterPriority;
  external_identifier: string | null;
  discharge_summary_advice: string | null;
}

export interface EncounterListRead extends EncounterBase {
  id: string;
  patient: PatientListRead;
  facility: FacilityBareMinimum;
  status_history: StatusHistory;
  encounter_class_history: EncounterClassHistory;
  created_date: string;
  modified_date: string;
  tags: TagConfig[];
  current_location: LocationMinSpec | null;
  care_team: CareTeamResponse[];
}

export interface EncounterRead
  extends Omit<EncounterListRead, "patient">, Permissions {
  appointment: AppointmentRead | null;
  patient: PatientRead;
  organizations: FacilityOrganizationRead[];
  current_location: LocationRead | null;
  location_history: LocationAssociationRead[];
  care_team: CareTeamResponse[];
  tags: TagConfig[];
  created_by: UserReadMinimal;
  updated_by: UserReadMinimal;
}
export interface EncounterCreate extends Omit<
  EncounterBase,
  "discharge_summary_advice" | "external_identifier"
> {
  patient: string;
  facility: string;
  organizations: string[];
  tags?: string[];
  appointment?: string;
  external_identifier?: string;
  discharge_summary_advice?: string;
}

export type EncounterEdit = EncounterBase;

export const completedEncounterStatus = ["completed"];
export const inactiveEncounterStatus = [
  ...["cancelled", "entered_in_error", "discontinued"],
  ...completedEncounterStatus,
] as const;

export const ENCOUNTER_STATUS_FILTER_COLORS = {
  planned: "bg-blue-100 text-blue-900 border-blue-300",
  in_progress: "bg-yellow-100/80 text-yellow-900 border-yellow-300",
  on_hold: "bg-orange-100 text-orange-900 border-orange-300",
  discharged: "bg-primary-100 text-primary-900 border-primary-300",
  completed: "bg-green-100 text-green-900 border-green-300",
  cancelled: "bg-red-100 text-red-900 border-red-300",
  discontinued: "bg-red-100 text-red-900 border-red-300",
  entered_in_error: "bg-red-100 text-red-900 border-red-300",
  unknown: "bg-gray-100 text-gray-900 border-gray-300",
} as const;

export const ENCOUNTER_CLASS_FILTER_COLORS = {
  imp: "bg-indigo-100 text-indigo-900 border-indigo-300",
  emer: "bg-red-100 text-red-900 border-red-300",
  amb: "bg-green-100 text-green-900 border-green-300",
  obsenc: "border-gray-300 bg-gray-100 text-gray-900",
  vr: "border-gray-300 bg-gray-100 text-gray-900",
  hh: "bg-teal-100 text-teal-900 border-teal-300",
} as const;

export const ENCOUNTER_PRIORITY_FILTER_COLORS = {
  stat: "bg-red-100 text-red-900 border-red-300",
  ASAP: "bg-yellow-100/80 text-yellow-900 border-yellow-300",
  emergency: "bg-red-100 text-red-900 border-red-300",
  urgent: "bg-orange-100 text-orange-900 border-orange-300",
  routine: "bg-blue-100 text-blue-900 border-blue-300",
  elective: "bg-indigo-100 text-indigo-900 border-indigo-300",
  rush_reporting: "bg-orange-100 text-orange-900 border-orange-300",
  timing_critical: "bg-yellow-100/80 text-yellow-900 border-yellow-300",
  callback_results: "bg-green-100 text-green-900 border-green-300",
  callback_for_scheduling: "bg-purple-100 text-purple-900 border-purple-300",
  preop: "bg-pink-100 text-pink-900 border-pink-300",
  as_needed: "bg-teal-100 text-teal-900 border-teal-300",
  use_as_directed: "bg-indigo-100 text-indigo-900 border-indigo-300",
} as const;
